src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,403 @@
+     */
+    protected int reverseMul;
+    protected FieldValueHitQueue<FieldValueHitQueue.Entry>[] compoundScores;
+    protected boolean queueFull[];
+    @Getter
+    protected float maxScore = 0.0f;
+    protected int[] collectedHits;
+    /*
+       searchSortPartOfIndexSort is used to evaluate whether to perform index sort or not.
+     */
+    private Boolean searchSortPartOfIndexSort = null;
+
+    private static final TopFieldDocs EMPTY_TOPDOCS = new TopFieldDocs(
+        new TotalHits(0, TotalHits.Relation.EQUAL_TO),
+        new ScoreDoc[0],
+        new SortField[0]
+    );
+
+    // Declaring the constructor private prevents extending this class by anyone
+    // else. Note that the class cannot be final since it's extended by the
+    // internal versions. If someone will define a constructor with any other
+    // visibility, then anyone will be able to extend the class, which is not what
+    // we want.
+    HybridTopFieldDocSortCollector(final int numHits, final HitsThresholdChecker hitsThresholdChecker, final Sort sort) {
+        this.numHits = numHits;
+        this.hitsThresholdChecker = hitsThresholdChecker;
+        this.sort = sort;
+    }
+
+    /**
+     * HybridCollectorManager fetches the topDocs in the reduce method.
+     * @return List of TopFieldDocs which represents results of Top Docs of individual subquery.
+     */
+    public List<TopFieldDocs> topDocs() {
+        if (compoundScores == null) {
+            return new ArrayList<>();
+        }
+
+        List<TopFieldDocs> topFieldDocs = new ArrayList<>();
+        for (int subQueryNumber = 0; subQueryNumber < compoundScores.length; subQueryNumber++) {
+            topFieldDocs.add(
+                topDocsPerQuery(
+                    0,
+                    Math.min(collectedHits[subQueryNumber], compoundScores[subQueryNumber].size()),
+                    compoundScores[subQueryNumber],
+                    collectedHits[subQueryNumber],
+                    sort.getSort()
+                )
+            );
+        }
+        return topFieldDocs;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return hitsThresholdChecker.scoreMode();
+    }
+
+    protected abstract class HybridTopDocSortLeafCollector implements LeafCollector {
+        protected HybridQueryScorer compoundQueryScorer;
+        private boolean collectedAllCompetitiveHits = false;
+        @Nullable
+        private FieldDoc after;
+        private final Sort sort;
+        private boolean initializePerSegment;
+
+        public HybridTopDocSortLeafCollector(Sort sort, @Nullable FieldDoc after) {
+            this.sort = sort;
+            this.after = after;
+            this.initializePerSegment = true;
+        }
+
+        @Override
+        public void setScorer(final Scorable scorer) throws IOException {
+            if (scorer instanceof HybridQueryScorer) {
```

Comment:
What if Scorer is null !! , it is going to break the flow.. Can we check NPE here

Reply:
I am bit confused on what we are checking here , 

If you see if Statement , it checks it is instance of HybridQueryScorer and again in else also we are checking the same thing in different function getHybridQueryScorer , 

if (scorer == null) {
                return null;
            }
            if (scorer instanceof HybridQueryScorer) {
                return (HybridQueryScorer) scorer;
            }
            
            Can we remove this multiple if else and reduce cyclometric complexicity of this code . 
            
            and It's fundamentally wrong becuase same check is happening in if and else.

Reply:
No, it is not same. The scorer can be nested.

Reply:
This call is made from core that too at Shard level.